lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts

```diff
@@ -0,0 +1,150 @@
+        headers[HeaderNames.CONTENT_TYPE] = Constants.URL_FORM_CONTENT_TYPE;
+        const networkRequestOptions: NetworkRequestOptions = { headers };
+        if (managedIdentityRequest.forceRefresh) {
+            networkRequestOptions.body =
+                networkRequest.computeParametersBodyString();
+        }
+
+        const reqTimestamp = TimeUtils.nowSeconds();
+        let serverTokenResponse: ServerAuthorizationTokenResponse;
+        let response;
+        try {
+            // Sources that send POST requests: Cloud Shell
+            if (networkRequest.httpMethod === HttpMethod.POST) {
+                response =
+                    await this.networkClient.sendPostRequestAsync<ServerAuthorizationTokenResponse>(
+                        networkRequest.computeUri(),
+                        networkRequestOptions
+                    );
+                // Sources that send GET requests: App Service, Azure Arc, IMDS, Service Fabric
+            } else {
+                response =
+                    await this.networkClient.sendGetRequestAsync<ServerAuthorizationTokenResponse>(
+                        networkRequest.computeUri(),
+                        networkRequestOptions
+                    );
+            }
+
+            serverTokenResponse = response.body;
+            serverTokenResponse.status = response.status;
+            // if success
+            if (serverTokenResponse.expires_on) {
+                serverTokenResponse.expires_in = serverTokenResponse.expires_on;
+            }
+            // if error
+            if (serverTokenResponse.message) {
+                serverTokenResponse.error = serverTokenResponse.message;
+            }
+            // if error
+            if (serverTokenResponse.correlationId) {
+                serverTokenResponse.correlation_id =
+                    serverTokenResponse.correlationId;
+            }
+        } catch (error) {
+            if (error instanceof AuthError) {
+                throw error;
+            } else {
+                throw createClientAuthError(ClientAuthErrorCodes.networkError);
+            }
+        }
+
+        const responseHandler = new ResponseHandler(
+            managedIdentityId.getId,
+            this.cacheManager,
+            this.cryptoProvider,
+            this.logger,
+            null,
+            null
+        );
+
+        responseHandler.validateTokenResponse(
+            serverTokenResponse,
+            refreshAccessToken
+        );
+
+        // caches the token
+        const authResult = await responseHandler.handleServerTokenResponse(
+            serverTokenResponse,
+            fakeAuthority,
+            reqTimestamp,
+            managedIdentityRequest
+        );
+
+        return ManagedIdentityUtils.convertAuthResultToManagedIdentityResult(
+            authResult
+        );
+    }
+
+    // used in unit tests
+    public isAppService(): boolean {
```

Comment:
This is a code smell. Base class should not know about derived class. 

Why don't you just copy this logic to the tests?